pkg/cmd/dev/ui.go

```diff
@@ -248,3 +247,205 @@ Replaces 'make ui-lint'.`,
 
 	return lintCmd
 }
+
+// arrangeFilesForTestWatchers moves files from Bazel's build output directory into the locations they'd be found during
+// a non-Bazel build, so that test watchers can successfully operate outside of the bazel sandbox.
+func arrangeFilesForTestWatchers(d *dev) error {
+	bazelBin, err := d.getBazelBin(d.cli.Context())
+	if err != nil {
+		return err
+	}
+
+	ossProtobufSrc := filepath.Join(bazelBin, "pkg", "ui", "workspaces", "db-console")
+	cclProtobufSrc := filepath.Join(ossProtobufSrc, "ccl")
+
+	dstDirs, err := getUIDirs(d)
+	if err != nil {
+		return err
+	}
+	ossProtobufDst := dstDirs.dbConsole
+	cclProtobufDst := filepath.Join(ossProtobufDst, "ccl")
+
+	protoFiles := []string{
+		filepath.Join("src", "js", "protos.js"),
+		filepath.Join("src", "js", "protos.d.ts"),
+	}
+
+	// Delete and recreate protobuf client files that were previously copied out of the sandbox
+	for _, relPath := range protoFiles {
+		ossDst := filepath.Join(ossProtobufDst, relPath)
+		cclDst := filepath.Join(cclProtobufDst, relPath)
+		if err := d.os.Remove(ossDst); err != nil {
+			return err
+		}
+		if err := d.os.Remove(cclDst); err != nil {
+			return err
+		}
+
+		if err := d.os.CopyFile(filepath.Join(ossProtobufSrc, relPath), ossDst); err != nil {
+			return err
+		}
+		if err := d.os.CopyFile(filepath.Join(cclProtobufSrc, relPath), cclDst); err != nil {
+			return err
+		}
+	}
+
+	err = d.os.RemoveAll(filepath.Join(dstDirs.clusterUI, "dist"))
+	if err != nil {
+		return err
+	}
+
+	err = d.os.CopyAll(
+		filepath.Join(bazelBin, "pkg", "ui", "workspaces", "cluster-ui", "dist"),
+		filepath.Join(dstDirs.clusterUI, "dist"),
+	)
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// makeUIWatchCmd initializes the 'ui test' subcommand, which runs unit tests for both db-console and cluster-ui.
+func makeUITestCmd(d *dev) *cobra.Command {
+	const (
+		// watchFlag is the name of the boolean long (GNU-style) flag that determines whether tests should be re-run when
+		// source files and test files change
+		watchFlag        = "watch"
+		verboseFlag      = "verbose"
+		streamOutputFlag = "stream-output"
+	)
+
+	testCmd := &cobra.Command{
+		Use:   "test",
+		Short: "Runs tests for UI subprojects",
+		Long: `Runs unit tests for db-console and cluster-ui, optionally setting up watchers to rerun those tests when files change.
+
+Replaces 'make ui-test' and 'make ui-test-watch'.`,
+		Args: cobra.MinimumNArgs(0),
+		RunE: func(cmd *cobra.Command, commandLine []string) error {
+			isDebug, _ := cmd.PersistentFlags().GetBool("debug")
+
+			// Create a context that cancels when OS signals come in
+			ctx, stop := signal.NotifyContext(d.cli.Context(), os.Interrupt, os.Kill)
+			defer stop()
+
+			isWatch, err := cmd.Flags().GetBool(watchFlag)
```

Comment:
Nit: you can use the helper function `mustGetFlagBool()` to get rid of this repetitive error-checking.